File: reducer.rb
#!/usr/bin/env ruby

require 'mongo'
require 'v8'

include Mongo

@con = MongoClient.new("localhost").db("tesis")

def process
  while(true)
    pendientes = @con["workers"].find({"status" => "reduce_pending"}).to_a
    if(pendientes.empty?)
      sleep(30)
      next
    end

    pendientes.each do |worker|
      process_pending(worker) 
    end
  end
end

def pre_reduce_map(map_results)
 #{1 => {"llave" : ["1","1","4"]}, 2 =>{"llave" : ["9"]}} => {"llave" => ["1","1","4","9"]}

 h = {}
 map_results.each do |k, v|
   v.each do |key, value|
     if(!h[key])
       h[key] = value
     else
       h[key] = h[key].concat(value)
     end
   end    
 end

 return h
end

def process_pending(worker)
  cxt = V8::Context.new
  #[{"llave"=>["1", "1", "4"]}]
  map_results = pre_reduce_map(worker["map_results"])
  map_results.each do |key, value|
    cxt["k"] = key
    cxt["v"] = value
    worker["reduce_results"][key] = cxt.eval("investigador_reduce = " + worker["reduce"] + ";investigador_reduce(k, v)");
  end
  @con["workers"].update({"_id" => worker["_id"]}, worker)
  worker["map_results"] = []
  worker["status"] = "finished"
  worker["data"] = []
  @con["workers"].update({"_id" => worker["_id"]}, worker)
end

process

File: reducer.js
(function() {
  var MongoClient, assert, db_url, sleep;

  db_url = 'mongodb://127.0.0.1:27017/tesis';

  MongoClient = require('mongodb').MongoClient;

  sleep = require('sleep');

  assert = require('assert');

  MongoClient.connect(db_url, function(err, connection) {
    var pre_reduce_map, process, process_pending;
    (process = function() {
      return connection.collection("workers", function(err, collection) {
        return collection.find({
          "status": "reduce_pending"
        }).toArray(function(err, pendientes) {
          var worker, _i, _len;
          if (pendientes.length === 0) {
            sleep.sleep(30);
          } else {
            for (_i = 0, _len = pendientes.length; _i < _len; _i++) {
              worker = pendientes[_i];
              process_pending(worker);
            }
          }
          return process();
        });
      });
    })();
    process_pending = function(worker) {
      var k, key, map_results, v, value;
      map_results = pre_reduce_map(worker.map_results);
      for (key in map_results) {
        value = map_results[key];
        k = key;
        v = value;
        worker["reduce_results"][key] = eval("investigador_reduce = " + worker["reduce"] + ";investigador_reduce(k, v)");
      }
      return connection.collection("workers", function(err, collection) {
        return collection.update({
          _id: worker._id
        }, {
          $set: {
            reduce_results: worker.reduce_results,
            map_results: [],
            status: "finished",
            data: []
          }
        }, function(err, count) {
          return assert.equal(1, count);
        });
      });
    };
    return pre_reduce_map = function(map_results) {
      var h, k, key, v, value;
      h = {};
      for (k in map_results) {
        v = map_results[k];
        for (key in v) {
          value = v[key];
          if (!h[key]) {
            h[key] = value;
          } else {
            h[key] = h[key].concat(value);
          }
        }
      }
      return h;
    };
  });

}).call(this);
